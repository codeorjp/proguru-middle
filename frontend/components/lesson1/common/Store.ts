import { action, decorate, observable, toJS } from "mobx";
import Storage from "components/lesson1/common/util/Storage";
import { DesignSchemaType } from "common/types/DesignSchemaType";
import { ComponentStyleType } from "common/types/ComponentStyleType";
import { LessonType } from "common/types/LessonType";
import { StageType } from "common/types/StageType";

type ClickElement = {
  title: string;
  className: string;
  style: ComponentStyleType;
  editDesignComponent: string[];
};

type StageConfig = {
  className: string;
  editDesignComponent: string[];
}[];

class Store {
  storage: Storage;
  schemas: DesignSchemaType;
  clickElm: ClickElement;
  stageConfig: StageConfig;
  sendSpeechText: string;
  receiveSpeechText: string;
  isFirstTime: boolean;
  isStageFirstTime: boolean;

  constructor(
    public userId: number | string,
    public lesson: LessonType,
    public stage: StageType,
    public submittedWorkspace: any
  ) {
    this.storage = new Storage(`${window.location.href}-user${this.userId}`);
    this.schemas = {
      SendSpeechBalloon: {
        style: {
          font: { fontSize: "16px", textAlign: "right" },
          color: "#000",
          border: {},
          background: "",
          borderRadius: "",
        },
      },
      SendProfileIcon: {
        style: {
          imgSize: {
            width: "200px",
            height: "200px",
          },
          border: {
            borderWidth: "1px",
          },
          borderRadius: "",
          background: "#fff",
        },
      },
      ReceiveSpeechBalloon: {
        style: {
          font: { fontSize: "16px", textAlign: "right" },
          color: "#000",
          border: {},
          borderRadius: "",
          background: "",
        },
      },
      ReceiveProfileIcon: {
        style: {
          imgSize: {
            width: "200px",
            height: "200px",
          },
          border: {
            borderWidth: "1px",
          },
          borderRadius: "",
          background: "#fff",
        },
      },
      ChatInput: {
        style: {
          font: { fontSize: "14px" },
          color: "#000",
          border: {},
          borderRadius: "",
          background: "#fff",
          padding: "",
        },
      },
      SubmitButton: {
        style: {
          font: { fontSize: "14px" },
          color: "#000",
          border: {},
          borderRadius: "",
          background: "#fff",
          padding: "",
        },
      },
      ChatBody: {
        style: {
          background: "#fff",
        },
      },
    };
    this.clickElm = {
      title: "", // 編集するコンポーネントタイトル
      className: "", // CSSクラス名
      style: {}, // 当てるスタイル
      editDesignComponent: [], // 編集コンポーネント
    };
    this.stageConfig = [
      {
        // アタッチするデザイン編集項目
        className: "", // CSSクラス名
        editDesignComponent: [], // 編集コンポーネント
      },
    ];
    this.sendSpeechText = "こんにちは，私の名前は美咲です。";
    this.receiveSpeechText = "こんにちは，私の名前は翔太です。";
    this.isFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}:isFirstTime`
    );
    this.isStageFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}-stage${this.stage.number}`
    );

    // sessionStorageのrestore処理
    this.schemas = this.storage.restoreWorkspace(this.schemas);
  }

  restoreSubmittedWorkspace(schemas: DesignSchemaType): void {
    if (schemas) {
      this.schemas = schemas;
    }
  }

  checkFirstTime(url: string): boolean {
    try {
      const isFirstTime = window.localStorage.getItem(url);
      return isFirstTime !== "false";
    } catch {
      return true;
    }
  }

  setFirstTimeDone(url: string): void {
    this.isFirstTime = false;
    try {
      window.localStorage.setItem(url, "false");
    } catch {}
  }

  setClickElm(clickElm: ClickElement): void {
    this.clickElm = clickElm;
  }

  setStyle(className: string, style: ComponentStyleType): void {
    const prevStyle = toJS(this.schemas[className].style);
    const nextStyle = {
      ...prevStyle,
      ...style,
    };
    this.schemas[className].style = nextStyle;
    this.clickElm.style = nextStyle;

    this.storage.saveWorkspace(this.schemas);
  }

  setSendSpeechText(text: string): void {
    this.sendSpeechText = text;
  }

  setReceiveSpeechText(text: string): void {
    this.receiveSpeechText = text;
  }

  setStageConfig(config: StageConfig): void {
    this.stageConfig = config;
  }
}

decorate(Store, {
  clickElm: observable,
  schemas: observable,
  isFirstTime: observable,
  sendSpeechText: observable,
  receiveSpeechText: observable,
  setClickElm: action,
  setStageConfig: action,
  setStyle: action,
  setSendSpeechText: action,
  setReceiveSpeechText: action,
  setFirstTimeDone: action,
});

export default Store;
