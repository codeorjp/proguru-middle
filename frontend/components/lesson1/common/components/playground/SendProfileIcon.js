import BaseProfileIcon from "components/lesson1/common/components/playground/baseComponent/BaseProfileIcon";
import { inject, observer } from "mobx-react";
import misaki from "common/statics/misaki.png";

class SendProfileIcon extends BaseProfileIcon {
  constructor(props) {
    super(props);
    this.componentName = "SendProfileIcon";
    this.componentTitle = "メッセージ送信側のアイコン";
    this.icon = misaki;
  }
}

SendProfileIcon.defaultProps = {
  isSample: false,
};

export default inject("store")(observer(SendProfileIcon));
