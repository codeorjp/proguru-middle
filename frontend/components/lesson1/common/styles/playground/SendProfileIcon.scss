
.iconImage {
  width: 100%;
  height: 100%;
}

.profileIcon {
  border-style: solid;
  border-color: #000;
  overflow: hidden;
}