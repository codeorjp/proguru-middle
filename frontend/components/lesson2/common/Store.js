import axios from "common/utils/axios";
import BlocklyJS from "blockly/javascript";
import moment from "moment";
import { decorate, observable } from "mobx";
import BlocklyLessonStore from "common/BlocklyLesson/BlocklyLessonStore";
import AnimationStore from "./AnimationStore";

class Store extends BlocklyLessonStore {
  constructor(userId, nickName, userIcon, lesson, stage, submittedWorkspace) {
    super(userId, submittedWorkspace);
    this.nickName = nickName;
    this.userIcon = userIcon;
    this.lesson = lesson;
    this.stage = stage;
    this.AnimationStore = AnimationStore;
    this.animate = this.animate.bind(this);
    this.receivedText = "";
    this.variables = {
      responseText: "",
      text: "",
    };
    this.servers = {
      serverA: "/api/say_hello",
      serverB: "cannot access serverB",
    };
    this.connectedServer = "";
    this.isFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}:isFirstTime`
    );
    this.isStageFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}-stage${this.stage.number}`
    );
  }

  init() {
    this.AnimationStore.reset();
    this.isExecute = false;
    this.receivedText = "";
    this.variables = {
      responseText: "",
      text: "",
    };
    this.connectedServer = "";
  }

  animate() {
    this.AnimationStore.animate();
  }

  addCommand(type, args, blockId) {
    this.AnimationStore.commands.push({
      type,
      args,
      blockId,
    });
  }

  skip(blockId) {
    this.addCommand("skip", null, blockId);
  }

  display(varName, blockId) {
    const msg = this.variables[varName];
    if (msg) {
      this.addCommand("display", { text: unescape(msg) }, blockId);
    }
  }

  assignText(text) {
    this.variables.text = text;
  }

  getMeridiem() {
    return moment().format("a");
  }

  connect(varName, callback, blockId, resolve) {
    if (varName && this.servers[varName]) {
      this.addCommand("connect", { before: "サーバと接続中..." }, blockId);

      if (varName === "serverA") {
        // serverAを選択してる場合のみ接続可能
        this.addCommand(
          "connectSuccess",
          { before: "接続成功しました！" },
          blockId
        );
        this.connectedServer = this.servers[varName];
        callback();
        resolve();
      } else {
        this.addCommand(
          "connectError",
          { before: "接続が失敗しました..." },
          blockId
        );
        resolve();
      }
    }
  }

  send(varName, blockId, resolve) {
    if (varName && this.variables[varName] && this.connectedServer) {
      this.addCommand(
        "request",
        { before: "メッセージを送信中...", after: "送信に成功しました！" },
        blockId
      );
      const formData = new FormData();
      formData.append("body", unescape(this.variables[varName]));
      formData.append("hour", new Date().getHours());
      axios.post(this.connectedServer, formData).then((res) => {
        this.receivedText = res.data.reply;
        resolve();
      });
    }
  }

  receiveResponse(callback, blockId) {
    if (this.receivedText !== "") {
      this.addCommand(
        "response",
        {
          before: "サーバからの応答を待ってます...",
          after: "応答を受け取りました！",
        },
        blockId
      );
      this.variables.responseText = this.receivedText;
      callback();
    }
  }
}

decorate(Store, {
  isFirstTime: observable,
});

export default Store;
