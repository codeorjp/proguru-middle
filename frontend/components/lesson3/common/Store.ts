import _ from "lodash";
import BlocklyJS from "blockly/javascript";
import moment from "moment";
import { action, decorate, observable } from "mobx";
import BlocklyLessonStore from "common/BlocklyLesson/BlocklyLessonStore";
import axios from "common/utils/axios";
import Endpoint from "common/utils/Endpoint";

class Store extends BlocklyLessonStore {
  nickName: string;
  studentNumber: number;
  userIcon: any; // blob?
  lesson: any;
  stage: any;
  submitCallback: Function | null;
  fetchServerID: number | null;
  isFirstTime: boolean;
  isStageFirstTime: boolean;
  messageBoardStyle: object;
  messages: object[];
  variables: any;
  last_id: number;

  constructor(
    userId,
    nickName,
    studentNumber,
    userIcon,
    lesson,
    stage,
    submittedWorkspace
  ) {
    super(userId, submittedWorkspace);
    this.nickName = nickName;
    this.studentNumber = studentNumber;
    this.userIcon = userIcon;
    this.lesson = lesson;
    this.stage = stage;
    this.fetchServerID = null;
    this.isFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}:isFirstTime`
    );
    this.isStageFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}-stage${this.stage.number}`
    );
    this.init();
    axios.get("/api/lessons/1/stages/8/workspaces").then((res) => {
      const { workspace } = res.data;
      this.messageBoardStyle = workspace ? JSON.parse(workspace.body) : null;
    });
  }

  init() {
    clearInterval(this.fetchServerID);
    this.isExecute = false;
    this.messages = [];
    this.submitCallback = async () => {};
    this.variables = {
      sendText: "",
      receivedText: "",
      newMessage: "",
      inputText: "",
      text: "",
    };
    this.last_id = null;
  }

  setChatInput(inputText: string) {
    this.variables.inputText = {
      messageId: _.uniqueId("message_not_sent_"),
      senderId: this.userId,
      senderType: "Student",
      userName: this.nickName,
      userNumber: this.studentNumber,
      icon: this.userIcon,
      content: inputText,
      createdAt: moment(),
      type: "Text",
    };
  }

  submit() {
    this.submitCallback().then((callback) => {
      if (callback) {
        callback(() => this.setChatInput(""));
      }
    });
  }

  setInputFileElement() {}

  // CommonToReceiveAndSend Function
  display(varName: string) {
    const message = this.variables[varName];
    if (message && message.content) {
      this.messages.push({
        ...message,
      });
    }
  }

  assignText(text: string) {
    this.variables.text = {
      messageId: _.uniqueId("message_not_sent_"),
      senderId: this.userId,
      senderType: "Student",
      userName: this.nickName,
      userNumber: this.studentNumber,
      icon: this.userIcon,
      content: unescape(text),
      createdAt: moment(),
      type: "Text",
    };
  }

  fetchMessage(callback: Function) {
    axios
      .get(Endpoint.message, {
        params: {
          kind: "lesson",
          last_id: this.last_id,
        },
      })
      .then((res) => {
        const { messages } = res.data;
        if (messages.length === 0) return;
        this.last_id = messages[messages.length - 1].id;

        _.map(messages, (message) => {
          const {
            id,
            sender_id,
            sender_type,
            nickname,
            number,
            icon_url,
            sent_at,
            body,
          } = message;
          this.variables.newMessage = {
            messageId: id,
            senderId: sender_id,
            senderType: sender_type,
            userName: nickname,
            userNumber: number,
            icon: icon_url,
            content: body,
            createdAt: moment(sent_at * 1000),
            type: "Text",
          };
          callback();
        });
      })
      .catch((error) => {
        console.log(error);
        clearInterval(this.fetchServerID);
      });
  }

  // Receive Function
  fetchServer(callback: Function) {
    this.fetchMessage(callback);
    this.fetchServerID = window.setInterval(() => {
      this.fetchMessage(callback);
    }, 5000);
  }

  isResponseIncludeText() {
    let isInclude = false;
    if (
      _.has(this.variables.newMessage, "content") &&
      this.variables.newMessage.content !== ""
    ) {
      isInclude = true;
    }
    return isInclude;
  }

  assignReceivedText(varName: string) {
    if (varName) {
      this.variables[varName] = this.variables.newMessage;
    }
  }

  // Send Function
  onSubmit(callback: Function) {
    this.submitCallback = callback;
  }

  send(varName: string, callback: Function, resolve: (value?: string) => void) {
    if (varName && this.variables[varName]) {
      if (this.variables[varName].content) {
        // テキストが空の場合は送信しない
        const params = new URLSearchParams();
        params.append("kind", "lesson");
        params.append("body", this.variables[varName].content);
        axios.post(Endpoint.message, params).then((res) => {
          this.variables[varName].messageId = res.data.id;
          this.variables[varName].createdAt = moment(res.data.created_at);
          callback();
          resolve();
        });
      }
    }
  }

  assignChatInput(varName: string) {
    if (varName) {
      this.variables[varName] = this.variables.inputText;
    }
  }
}

decorate(Store, {
  isFirstTime: observable,
  variables: observable,
  messages: observable,
  messageBoardStyle: observable,
  setChatInput: action,
  submit: action,
});

export default Store;
