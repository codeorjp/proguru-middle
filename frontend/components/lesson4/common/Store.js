/* eslint-disable camelcase */
/* eslint-disable prefer-destructuring */
import _ from "lodash";
import BlocklyJS from "blockly/javascript";
import moment from "moment";
import axios from "common/utils/axios";
import { action, decorate, observable } from "mobx";
import BlocklyLessonStore from "common/BlocklyLesson/BlocklyLessonStore";
import Endpoint from "common/utils/Endpoint";
import { ObjectURLToBlob } from "common/BlocklyLesson/utils/ImageTranslator";
import image1 from "./statics/image1.jpg";
import image2 from "./statics/image2.jpg";
import image3 from "./statics/image3.jpg";
import image4 from "./statics/image4.jpg";
import image5 from "./statics/image5.jpg";
import image6 from "./statics/image6.jpg";
import image7 from "./statics/image7.jpg";
import image8 from "./statics/image8.jpg";
import image9 from "./statics/image9.jpg";
import image10 from "./statics/image10.jpg";

const createObjectURL =
  (window.URL || window.webkitURL).createObjectURL || window.createObjectURL;
const sampleImages = {
  1: image1,
  2: image2,
  3: image3,
  4: image4,
  5: image5,
  6: image6,
  7: image7,
  8: image8,
  9: image9,
  10: image10,
};

class Store extends BlocklyLessonStore {
  constructor(
    userId,
    nickName,
    studentNumber,
    userIcon,
    lesson,
    stage,
    submittedWorkspace
  ) {
    super(userId, submittedWorkspace);
    this.nickName = nickName;
    this.studentNumber = studentNumber;
    this.userIcon = userIcon;
    this.lesson = lesson;
    this.stage = stage;
    this.fetchServerID = "";
    this.processID = "";
    this.isFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}:isFirstTime`
    );
    this.isStageFirstTime = this.checkFirstTime(
      `lesson${this.lesson.number}-stage${this.stage.number}`
    );
    this.setSampleImage = this.setSampleImage.bind(this);
    this.init();
    axios.get("/api/lessons/1/stages/8/workspaces").then((res) => {
      const { workspace } = res.data;
      this.messageBoardStyle = workspace ? JSON.parse(workspace.body) : null;
    });
  }

  init() {
    clearInterval(this.fetchServerID);
    clearInterval(this.processID);
    this.isExecute = false;
    this.messages = [];
    this.variables = {
      receivedImage: "",
      sendImage: "",
      newImage: "",
      selectedImage: "", // `選ばれた画像`ブロックの変数
      inputText: "",
    };
    this.imageQuality = 1.0;
    this.selectedImageFile = "";
    this.clickSelectImageCallback = () => {};
    this.callbackSelectImage = () => {};
    this.callbackSelectImagePromiseResolve = () => {};
    this.last_id = null;
    this.newMessages = [];
  }

  setSampleImage() {
    const property = {
      senderId: this.userId,
      senderType: "Student",
      userName: this.nickName,
      userNumber: this.studentNumber,
      icon: this.userIcon,
      createdAt: moment(),
      type: "Image",
    };

    Object.keys(sampleImages).forEach((key) => {
      this.variables[`sampleImage${key}`] = {
        ...property,
        messageId: _.uniqueId("message_not_sent_"),
        content: sampleImages[key],
      };
    });
  }

  // CommonToReceiveAndSend Function
  display(varName) {
    const message = this.variables[varName];
    if (message && message.content) {
      const { content } = message;
      this.messages.push({
        ...message,
        content:
          typeof content === "object" ? createObjectURL(content) : content,
      });
    }
  }

  getImageWidth(varName, resolve) {
    const msg = this.variables[varName];
    if (!msg) {
      resolve("no msg");
      return;
    }

    const img = new Image();
    img.src = msg.content;
    img.onload = () => resolve(img.width);
  }

  getImageHeight(varName, resolve) {
    const msg = this.variables[varName];
    if (!msg) {
      resolve("no msg");
      return;
    }

    const img = new Image();
    img.src = msg.content;
    img.onload = () => resolve(img.height);
  }

  compression(object, resolve) {
    const { varName, compressionRate } = object;
    const msg = this.variables[varName];
    if (!msg) {
      resolve("no msg");
      return;
    }

    this.imageQuality = compressionRate / 100;
    const imgSrc = msg.content;
    const img = new Image();
    const canvas = document.createElement("canvas");
    img.crossOrigin = "Anonymous";
    img.addEventListener("load", async () => {
      canvas.width = img.width;
      canvas.height = img.height;
      const ctx = canvas.getContext("2d");
      ctx.clearRect(0, 0, canvas.width, canvas.height);
      ctx.drawImage(img, 0, 0, canvas.width, canvas.height);

      this.variables[varName].content = await this._canvasToObjectURL(
        canvas,
        "image/jpeg"
      );
      resolve("compress");
    });
    if (!/blob/.test(imgSrc)) {
      const queryStringForHandleCORSError = "?not-from-cache-please";
      img.src = imgSrc + queryStringForHandleCORSError;
    } else {
      img.src = imgSrc;
    }
  }

  resizeImage(object, resolve) {
    const { varName, width, height } = object;
    const msg = this.variables[varName];
    if (!msg) {
      resolve("no msg");
      return;
    }

    this.imageQuality = 1.0;
    const imgSrc = msg.content;
    const img = new Image();
    const canvas = document.createElement("canvas");
    canvas.width = width;
    canvas.height = height;
    img.crossOrigin = "Anonymous";
    img.addEventListener("load", async () => {
      img.width = width;
      img.height = height;
      const ctx = canvas.getContext("2d");
      ctx.clearRect(0, 0, canvas.width, canvas.height);
      ctx.drawImage(img, 0, 0, width, height);

      this.variables[varName].content = await this._canvasToObjectURL(
        canvas,
        "image/jpeg"
      );
      resolve("resize");
    });
    if (!/blob/.test(imgSrc)) {
      const queryStringForHandleCORSError = "?not-from-cache-please";
      img.src = imgSrc + queryStringForHandleCORSError;
    } else {
      img.src = imgSrc;
    }
  }

  fixedAspectResize(object, resolve) {
    const { varName, isHeight, size } = object;
    const msg = this.variables[varName];
    if (!msg) {
      resolve("no msg");
      return;
    }

    this.imageQuality = 1.0;
    const imgSrc = msg.content;
    const img = new Image();
    const canvas = document.createElement("canvas");
    img.crossOrigin = "Anonymous";
    img.addEventListener("load", async () => {
      if (isHeight) {
        canvas.width = Math.round(img.width * (size / img.height));
        canvas.height = size;
      } else {
        canvas.height = Math.round(img.height * (size / img.width));
        canvas.width = size;
      }
      const ctx = canvas.getContext("2d");
      ctx.clearRect(0, 0, canvas.width, canvas.height);
      ctx.drawImage(img, 0, 0, canvas.width, canvas.height);

      this.variables[varName].content = await this._canvasToObjectURL(
        canvas,
        "image/jpeg"
      );
      resolve("resize");
    });
    if (!/blob/.test(imgSrc)) {
      const queryStringForHandleCORSError = "?not-from-cache-please";
      img.src = imgSrc + queryStringForHandleCORSError;
    } else {
      img.src = imgSrc;
    }
  }

  _canvasToObjectURL(canvas, type) {
    return new Promise((resolve) => {
      const base64 = canvas.toDataURL(type, this.imageQuality);
      const bin = atob(base64.replace(/^.*,/, ""));
      const buffer = new Uint8Array(bin.length);
      for (let i = 0; i < bin.length; i += 1) {
        buffer[i] = bin.charCodeAt(i);
      }
      const blob = new Blob([buffer.buffer], { type });
      return resolve(createObjectURL(blob));
    });
  }

  fetchMessage() {
    axios
      .get(Endpoint.message, {
        params: {
          kind: "lesson",
          last_id: this.last_id,
        },
      })
      .then((res) => {
        const { messages } = res.data;
        if (messages.length === 0) return;
        this.last_id = messages[messages.length - 1].id;

        _.map(messages, (message) => {
          const {
            id,
            sender_id,
            sender_type,
            nickname,
            number,
            icon_url,
            sent_at,
            image_url,
          } = message;
          if (image_url) {
            this.newMessages.push({
              messageId: id,
              senderId: sender_id,
              senderType: sender_type,
              userName: nickname,
              userNumber: number,
              icon: icon_url,
              content: image_url,
              createdAt: moment(sent_at * 1000),
              type: "Image",
            });
          }
        });
      })
      .catch((error) => {
        console.log(error);
        clearInterval(this.fetchServerID);
        clearInterval(this.processID);
      });
  }

  // Receive Function
  fetchServer(callback) {
    this.fetchMessage();
    this.fetchServerID = setInterval(() => {
      this.fetchMessage();
    }, 5000);

    this.processID = setInterval(() => {
      callback();
      this.newMessages.shift();
    }, 250);
  }

  isResponseIncludeImage() {
    return (
      _.has(this.newMessages[0], "content") &&
      this.newMessages[0].content !== ""
    );
  }

  assignReceivedImage(varName) {
    if (varName) {
      this.variables[varName] = this.newMessages[0];
    }
  }

  // SendImage Function
  onClickSelectImage(callback) {
    this.clickSelectImageCallback = callback;
  }

  clickImageButton() {
    this.clickSelectImageCallback();
  }

  setInputFileElement(element) {
    this.inputFileElement = element;
  }

  showFileDialog(callback, resolve) {
    this.callbackSelectImagePromiseResolve = resolve;
    this.callbackSelectImage = callback;
    this.inputFileElement.click();
  }

  onSelectImage(file) {
    // file: FileObject
    this.selectedImageFile = createObjectURL(file);
    this.variables.selectedImage = {
      messageId: _.uniqueId("message_not_sent_"),
      senderId: this.userId,
      senderType: "Student",
      userName: this.nickName,
      userNumber: this.studentNumber,
      icon: this.userIcon,
      content: this.selectedImageFile,
      createdAt: moment(),
      type: "Image",
    };
    this.callbackSelectImage(this.callbackSelectImagePromiseResolve);
  }

  async send(varName, callback) {
    if (varName && this.variables[varName]) {
      if (this.variables[varName].content) {
        // 画像がない場合は送信しない
        const formData = new FormData();
        formData.append("kind", "lesson");
        if (typeof this.variables[varName].content === "string") {
          formData.append(
            "image",
            await ObjectURLToBlob(this.variables[varName].content)
          );
        } else {
          formData.append("image", this.variables[varName].content);
        }

        axios.post(Endpoint.message, formData).then((res) => {
          this.variables[varName].messageId = res.data.id;
          this.variables[varName].createdAt = moment(res.data.created_at);
          callback();
        });
      }
    }
  }

  assignSelectedImage(varName) {
    if (varName) {
      this.variables[varName] = this.variables.selectedImage;
    }
  }
}

decorate(Store, {
  isFirstTime: observable,
  messages: observable,
  messageBoardStyle: observable,
  clickImageButton: action,
  setInputFileElement: action,
  onSelectImage: action,
});

export default Store;
