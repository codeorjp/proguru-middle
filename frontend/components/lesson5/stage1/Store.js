import axios from "common/utils/axios";
import Lesson1Store from "components/lesson1/common/Store";

class Store extends Lesson1Store {
  constructor(userId, lesson, stage, submittedWorkspace) {
    super(userId, lesson, stage, submittedWorkspace);

    // 提出済みのデザインも SessionStorage に保存されているデザインもなければ 1-8 のデザインを取得する
    if (!this.submittedWorkspace && !this.storage.isExistsSavedWorkspace()) {
      axios.get("/api/lessons/1/stages/8/workspaces").then((res) => {
        const { workspace } = res.data;
        this.schemas = workspace ? JSON.parse(workspace.body) : this.schemas;
      });
    }
  }
}

export default Store;
