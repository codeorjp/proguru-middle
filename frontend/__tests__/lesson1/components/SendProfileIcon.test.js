/* eslint-disable no-undef */
import React from "react";
import { Provider } from "mobx-react";
import renderer from "react-test-renderer";
import Store from "components/lesson1/common/Store";
import SendProfileIcon from "components/lesson1/common/components/playground/SendProfileIcon";

let store;
let component;

beforeEach(() => {
  // Storeのインスタンスを生成
  store = new Store(
    0, // userId
    { number: 0 }, // lesson
    { number: 0 }, // stage
    null // submittedWorkspace
  );
  store.setStyle("SendProfileIcon", {});

  component = renderer.create(
    <Provider store={store}>
      <SendProfileIcon />
    </Provider>
  );
});

describe("Send Profile Icon", () => {
  test("render SendProfileIcon", () => {
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("set width: 70px, height: 70px", () => {
    store.setStyle("SendProfileIcon", {
      imgSize: {
        width: "70px",
        height: "70px",
      },
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("set border-color: #A5A5A5, border-style: solid, boder-width: 1px", () => {
    store.setStyle("SendProfileIcon", {
      border: {
        borderColor: "#A5A5A5",
        borderStyle: "solid",
        borderWidth: "1px",
      },
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("set border-radius: 50px 50px 50px 50px", () => {
    store.setStyle("SendProfileIcon", {
      borderRadius: "50px 50px 50px 50px",
    });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("set background: #00ff00", () => {
    store.setStyle("SendProfileIcon", { background: "#00ff00" });

    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
