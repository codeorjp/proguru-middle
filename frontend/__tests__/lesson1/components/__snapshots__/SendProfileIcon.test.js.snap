// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Send Profile Icon render SendProfileIcon 1`] = `
<div
  className="undefined undefined"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    style={
      Object {
        "background": "#fff",
        "borderRadius": "",
        "borderWidth": "1px",
        "height": "200px",
        "width": "200px",
      }
    }
  >
    <img
      alt=""
      src="test-file-stub"
    />
  </div>
  <div
    className="
                undefined
                null
                undefined
              "
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;

exports[`Send Profile Icon set background: #00ff00 1`] = `
<div
  className="undefined undefined"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    style={
      Object {
        "background": "#00ff00",
        "borderRadius": "",
        "borderWidth": "1px",
        "height": "200px",
        "width": "200px",
      }
    }
  >
    <img
      alt=""
      src="test-file-stub"
    />
  </div>
  <div
    className="
                undefined
                null
                undefined
              "
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;

exports[`Send Profile Icon set border-color: #A5A5A5, border-style: solid, boder-width: 1px 1`] = `
<div
  className="undefined undefined"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    style={
      Object {
        "background": "#fff",
        "borderColor": "#A5A5A5",
        "borderRadius": "",
        "borderStyle": "solid",
        "borderWidth": "1px",
        "height": "200px",
        "width": "200px",
      }
    }
  >
    <img
      alt=""
      src="test-file-stub"
    />
  </div>
  <div
    className="
                undefined
                null
                undefined
              "
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;

exports[`Send Profile Icon set border-radius: 50px 50px 50px 50px 1`] = `
<div
  className="undefined undefined"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    style={
      Object {
        "background": "#fff",
        "borderRadius": "50px 50px 50px 50px",
        "borderWidth": "1px",
        "height": "200px",
        "width": "200px",
      }
    }
  >
    <img
      alt=""
      src="test-file-stub"
    />
  </div>
  <div
    className="
                undefined
                null
                undefined
              "
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;

exports[`Send Profile Icon set width: 70px, height: 70px 1`] = `
<div
  className="undefined undefined"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    style={
      Object {
        "background": "#fff",
        "borderRadius": "",
        "borderWidth": "1px",
        "height": "70px",
        "width": "70px",
      }
    }
  >
    <img
      alt=""
      src="test-file-stub"
    />
  </div>
  <div
    className="
                undefined
                null
                undefined
              "
    onMouseOut={[Function]}
    onMouseOver={[Function]}
  />
</div>
`;
